dali/test/python/test_operator_preemph.py

```diff
@@ -17,54 +17,83 @@
 import nvidia.dali.ops as ops
 import nvidia.dali.types as types
 import numpy as np
+from functools import partial
+from test_utils import compare_pipelines
+from test_utils import RandomlyShapedDataIterator
 
-sample_size = 100
-premade_batch = [np.random.rand(sample_size)]
-
-
-def isclose(a, b, rel_tol=1e-09, abs_tol=0.0):
-    """
-    Python2.7 doesn't have math.isclose()
-    """
-    return abs(a - b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)
-
-
-def preemph(signal, coeff=0.0):
-    ret = np.copy(signal)
-    if coeff == 0.0:
-        return ret
-    for i in range(sample_size - 1, 0, -1):
-        ret[i] -= coeff * ret[i - 1]
-    ret[0] -= coeff * ret[0]
-    return ret
+SEED = 12345
 
+def preemph_func(coeff, signal):
+    in_shape = signal.shape
+    assert(len(in_shape) == 1)  # 1D
+    out = np.copy(signal)
+    out[0]  -= coeff * signal[0]
+    out[1:] -= coeff * signal[0:in_shape[0]-1]
+    return out
 
 class PreemphasisPipeline(Pipeline):
-    def __init__(self, preemph_coeff=0., num_threads=1):
-        super(PreemphasisPipeline, self).__init__(1, num_threads, 0)
+    def __init__(self, device, batch_size, iterator, preemph_coeff=0.97, per_sample_coeff=False, num_threads=4, device_id=0):
+        super(PreemphasisPipeline, self).__init__(batch_size, num_threads, device_id, seed=SEED)
+        self.device = device
+        self.iterator = iterator
         self.ext_src = ops.ExternalSource()
-        self.preemph = ops.PreemphasisFilter(preemph_coeff=preemph_coeff, device="cpu")
+        self.per_sample_coeff = per_sample_coeff
+        self.uniform = ops.Uniform(range=(0.5, 0.97), seed=1234)
+        if self.per_sample_coeff:
+            self.preemph = ops.PreemphasisFilter(device=device)
+        else:
+            self.preemph = ops.PreemphasisFilter(device=device, preemph_coeff=preemph_coeff)
 
     def define_graph(self):
         self.data = self.ext_src()
-        return self.preemph(self.data)
+        out = self.data.gpu() if self.device == 'gpu' else self.data
+        if self.per_sample_coeff:
+            preemph_coeff_arg = self.uniform()
+            return self.preemph(out, preemph_coeff=preemph_coeff_arg)
+        else:
+            return self.preemph(out)
 
     def iter_setup(self):
```

Comment:
Can you rework to use the new ExternalSource API?

Reply:
OK